src/ScopeProvider/ScopeProvider.tsx

```diff
@@ -0,0 +1,111 @@
+import { Provider, useStore } from 'jotai/react';
+import {
+  createContext,
+  useContext,
+  useState,
+  type PropsWithChildren,
+} from 'react';
+import { createScope, type Scope } from './scope';
+import type { AnyAtom, Store } from './types';
+
+const ScopeContext = createContext<{
+  scope: Scope | undefined;
+  baseStore: Store | undefined;
+}>({ scope: undefined, baseStore: undefined });
+
+export const ScopeProvider = ({
+  atoms,
+  children,
+  debugName,
+}: PropsWithChildren<{ atoms: Iterable<AnyAtom>; debugName?: string }>) => {
+  const parentStore: Store = useStore();
+  let { scope: parentScope, baseStore = parentStore } =
+    useContext(ScopeContext);
+  // if this scope is the first descendant scope under Provider then we don't want to inherit parentScope
+  // https://github.com/jotaijs/jotai-scope/pull/33#discussion_r1604268003
+  if (isTopLevelScope(parentStore)) {
+    parentScope = undefined;
+    baseStore = parentStore;
+  }
+
+  /**
+   * atomSet is used to detect if the atoms prop has changed.
+   */
+  const atomSet = new Set(atoms);
+
+  function initialize() {
+    const scope = createScope(atoms, parentScope, debugName);
+
+    /**
+     * When an atom is accessed via useAtomValue/useSetAtom, the access should
+     * be handled by a router atom copy.
+     */
+    const patchedStore: PatchedStore = {
```

Comment:
btw, compatibility with `jotai-devtools` will be hard. I think there's an open issue.